File: nodejs-compressor/src/parser/nodes/TabularArrayNode.js
/**
 * @fileoverview Tabular Array AST Node for ASON 2.0
 *
 * Represents a tabular array with schema definition.
 * Format: @section [N]{field1,field2,...}
 *
 * @module TabularArrayNode
 * @license MIT
 * @version 2.0.0
 */

import { ASTNode, ArrayNode, ObjectNode } from './ASTNode.js';

/**
 * AST node representing a tabular array.
 *
 * Tabular arrays are compact representations of uniform data:
 * @users [2]{id,name,email}
 * 1|John|john@ex.com
 * 2|Jane|jane@ex.com
 *
 * @class TabularArrayNode
 * @extends ArrayNode
 */
export class TabularArrayNode extends ArrayNode {
  /**
   * Creates a tabular array node.
   *
   * @constructor
   * @param {string[]} schema - Field names (column schema)
   * @param {ObjectNode[]} rows - Array of row objects
   * @param {Object} [metadata={}] - Optional metadata
   *
   * @example
   * const tabular = new TabularArrayNode(
   *   ['id', 'name'],
   *   [
   *     new ObjectNode(new Map([['id', new PrimitiveNode(1)], ['name', new PrimitiveNode('John')]]))
   *   ]
   * );
   */
  constructor(schema = [], rows = [], metadata = {}) {
    super(rows, metadata);
    this.type = 'TabularArray';

    /** @type {string[]} Field names (schema) */
    this.schema = schema;

    /** @type {number|null} Expected row count (from [N] annotation) */
    this.expectedCount = metadata.expectedCount ?? null;
  }

  accept(visitor) {
    return visitor.visitTabularArray?.(this) ?? this.toValue();
  }

  /**
   * Validates that all rows conform to the schema.
   *
   * @returns {Object} Validation result
   * @returns {boolean} .valid - True if all rows are valid
   * @returns {string[]} .errors - Array of error messages
   */
  validate() {
    const errors = [];

    // Check row count if specified
    if (this.expectedCount !== null && this.elements.length !== this.expectedCount) {
      errors.push(
        `Expected ${this.expectedCount} rows, got ${this.elements.length}`
      );
    }

    // Check each row has all schema fields
    for (let i = 0; i < this.elements.length; i++) {
      const row = this.elements[i];

      if (!(row instanceof ObjectNode)) {
        errors.push(`Row ${i} is not an object`);
        continue;
      }

      // Check for missing required fields
      for (const field of this.schema) {
        if (!row.hasProperty(field)) {
          errors.push(`Row ${i} missing required field: ${field}`);
        }
      }

      // Warn about extra fields (not in schema)
      for (const [key] of row.properties) {
        if (!this.schema.includes(key)) {
          errors.push(`Row ${i} has unexpected field: ${key}`);
        }
      }
    }

    return {
      valid: errors.length === 0,
      errors
    };
  }

  /**
   * Checks if this tabular array has a uniform structure.
   *
   * @returns {boolean} True if all rows have exactly the schema fields
   */
  isUniform() {
    return this.elements.every(row =>
      row instanceof ObjectNode &&
      row.properties.size === this.schema.length &&
      this.schema.every(field => row.hasProperty(field))
    );
  }

  /**
   * Gets the actual row count.
   *
   * @returns {number} Number of rows
   */
  get rowCount() {
    return this.elements.length;
  }

  /**
   * Gets the field count (number of columns).
   *
   * @returns {number} Number of fields
   */
  get fieldCount() {
    return this.schema.length;
  }

  /**
   * Adds a row to the tabular array.
   *
   * @param {ObjectNode|Object} row - Row to add (ObjectNode or plain object)
   */
  addRow(row) {
    if (!(row instanceof ObjectNode)) {
      // Convert plain object to ObjectNode
      const objNode = new ObjectNode();
      for (const [key, value] of Object.entries(row)) {
        objNode.setProperty(key, value);
      }
      this.addElement(objNode);
    } else {
      this.addElement(row);
    }
  }

  /**
   * Gets a specific row.
   *
   * @param {number} index - Row index
   * @returns {ObjectNode|undefined} Row object or undefined
   */
  getRow(index) {
    return this.elements[index];
  }

  /**
   * Gets a specific cell value.
   *
   * @param {number} rowIndex - Row index
   * @param {string} field - Field name
   * @returns {*} Cell value or undefined
   */
  getCell(rowIndex, field) {
    const row = this.getRow(rowIndex);
    return row?.getProperty(field);
  }

  toString(indent = 0) {
    const spaces = '  '.repeat(indent);
    let str = `${spaces}TabularArray [${this.rowCount}]{${this.schema.join(',')}} [\n`;
    for (const row of this.elements) {
      str += row.toString(indent + 1) + '\n';
    }
    str += `${spaces}]`;
    return str;
  }

  toJSON() {
    return {
      ...super.toJSON(),
      schema: this.schema,
      expectedCount: this.expectedCount,
      rowCount: this.rowCount
    };
  }
}

File: python-compressor/src/ason/parser/tabular_array_node.py
"""
Tabular Array AST Node for ASON 2.0
"""

from typing import Any, Dict, List, Optional
from .ast_nodes import ArrayNode, ObjectNode


class TabularArrayNode(ArrayNode):
    """AST node representing a tabular array."""
    
    def __init__(self, schema: Optional[List[str]] = None, rows: Optional[List[ObjectNode]] = None, metadata: Optional[Dict[str, Any]] = None):
        super().__init__(rows or [], metadata)
        self.type = 'TabularArray'
        self.schema = schema or []
        self.expected_count = metadata.get('expectedCount') if metadata else None
    
    def validate(self) -> Dict[str, Any]:
        """Validate that all rows conform to the schema."""
        errors: List[str] = []
        
        # Check row count if specified
        if self.expected_count is not None and len(self.elements) != self.expected_count:
            errors.append(f"Expected {self.expected_count} rows, got {len(self.elements)}")
        
        # Check each row has all schema fields
        for i, row in enumerate(self.elements):
            if not isinstance(row, ObjectNode):
                errors.append(f"Row {i} is not an object")
                continue
            
            # Check for missing required fields
            for field in self.schema:
                if not row.has_property(field):
                    errors.append(f"Row {i} missing required field: {field}")
            
            # Warn about extra fields
            for key in row.properties.keys():
                if key not in self.schema:
                    errors.append(f"Row {i} has unexpected field: {key}")
        
        return {
            'valid': len(errors) == 0,
            'errors': errors
        }
    
    def is_uniform(self) -> bool:
        """Check if this tabular array has a uniform structure."""
        return all(
            isinstance(row, ObjectNode) and
            len(row.properties) == len(self.schema) and
            all(row.has_property(field) for field in self.schema)
            for row in self.elements
        )
    
    @property
    def row_count(self) -> int:
        return len(self.elements)
    
    @property
    def field_count(self) -> int:
        return len(self.schema)
    
    def add_row(self, row: Any) -> None:
        """Add a row to the tabular array."""
        if not isinstance(row, ObjectNode):
            obj_node = ObjectNode()
            for key, value in row.items():
                obj_node.set_property(key, value)
            self.add_element(obj_node)
        else:
            self.add_element(row)
    
    def get_row(self, index: int) -> Optional[ObjectNode]:
        """Get a specific row."""
        return self.elements[index] if 0 <= index < len(self.elements) else None
    
    def get_cell(self, row_index: int, field: str) -> Optional[Any]:
        """Get a specific cell value."""
        row = self.get_row(row_index)
        return row.get_property(field) if row else None
    
    def to_string(self, indent: int = 0) -> str:
        spaces = '  ' * indent
        result = f"{spaces}TabularArray [{self.row_count}]{{{','.join(self.schema)}}} [\n"
        for row in self.elements:
            result += row.to_string(indent + 1) + '\n'
        result += f"{spaces}]"
        return result
